lib/ddtrace/tracer.rb

```diff
@@ -63,61 +63,16 @@ def call_context(key = nil)
     # * +enabled+: set if the tracer submits or not spans to the local agent. It's enabled
     #   by default.
     def initialize(options = {})
-      # Configurable options
-      @context_flush = if options[:context_flush]
-                         options[:context_flush]
-                       elsif options[:partial_flush]
-                         Datadog::ContextFlush::Partial.new(options)
-                       else
-                         Datadog::ContextFlush::Finished.new
-                       end
-
+      @context_flush = options[:context_flush] || Datadog::ContextFlush::Finished.new
       @default_service = options[:default_service]
       @enabled = options.fetch(:enabled, true)
       @provider = options[:context_provider] || Datadog::DefaultContextProvider.new
-      @sampler = options.fetch(:sampler, Datadog::AllSampler.new)
-      @tags = options.fetch(:tags, {})
-      @writer = options.fetch(:writer) { Datadog::Writer.new }
-
-      # Instance variables
-      @mutex = Mutex.new
-
-      # Enable priority sampling by default
-      activate_priority_sampling!(@sampler)
-    end
-
-    # Updates the current \Tracer instance, so that the tracer can be configured after the
-    # initialization. Available +options+ are:
-    #
-    # * +enabled+: set if the tracer submits or not spans to the trace agent
-    # * +hostname+: change the location of the trace agent
-    # * +port+: change the port of the trace agent
-    # * +partial_flush+: enable partial trace flushing
-    #
-    # For instance, if the trace agent runs in a different location, just:
-    #
-    #   tracer.configure(hostname: 'agent.service.consul', port: '8777')
-    #
-    def configure(options = {})
-      enabled = options.fetch(:enabled, nil)
-
-      # Those are rare "power-user" options.
-      sampler = options.fetch(:sampler, nil)
-
-      @enabled = enabled unless enabled.nil?
-      @sampler = sampler unless sampler.nil?
-
-      configure_writer(options)
-
-      if options.key?(:context_flush) || options.key?(:partial_flush)
-        @context_flush = if options[:context_flush]
-                           options[:context_flush]
-                         elsif options[:partial_flush]
-                           Datadog::ContextFlush::Partial.new(options)
-                         else
-                           Datadog::ContextFlush::Finished.new
-                         end
-      end
+      @sampler = options[:sampler] || PrioritySampler.new(
+        base_sampler: Datadog::AllSampler.new,
+        post_sampler: Sampling::RuleSampler.new,
+      )
+      @tags = options[:tags] || {}
+      @writer = options[:writer] || Datadog::Writer.new
```

Comment:
Since this class is rather complex, I don't think we should try to somewhat duplicate its initialization logic here. E.g. `Tracer.new` should come from `#build_tracer` in the components; if we would like to have that be standalone, maybe `#build_tracer` should be moved to another class (?).

I don't see a very interesting use case for supporting `Tracer.new` and so having all these defaults seems a bit unneeded.

Reply:
@delner I'm torn on removing the ability to build a brand new `Datadog::Tracer` instance from scratch.

I've seen clients create custom tracers for testing, and they normally overwrite one method or two.